File: problems/#147/v2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* insertionSortList(ListNode* head) {
        ListNode* dummy = new ListNode(-5001);
        ListNode *pv, *curr;
        while (head != NULL) {
            curr = head;
            head = head->next;
            pv = dummy;
            while (pv->next != NULL && pv->next->val < curr->val) {
                pv = pv->next;
            }
            curr->next = pv->next == NULL ? NULL : pv->next;
            pv->next = curr;
        }
        return dummy->next;
    }
};

File: problems/#725/v2.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 func splitListToParts(head *ListNode, k int) []*ListNode {
    ret, cnt := make([]*ListNode, k), 0
    for curr := head; curr != nil; curr = curr.Next {
        cnt++
    }
    q, r := cnt / k, cnt % k
    for i := range k {
        ph, pt := head, &ListNode{0, head}
        for _ = range q {
            pt, head = pt.Next, head.Next
        }
        if r > 0 {
            pt, head = pt.Next, head.Next
            r--
        }
        pt.Next = nil
        ret[i] = ph
    }
    return ret
}